src/ui/wxWidgets/TreeCtrl.cpp

```diff
@@ -901,6 +1013,436 @@ void TreeCtrl::OnKeyDown(wxTreeEvent& evt)
+        m_drag_image->Show();
+      }
+    }
+    return;
+  }
+
+  evt.Skip();
+}
+
+void TreeCtrl::resetScrolling()
+{
+  StopAutoScrolling();
+  if(m_scroll_timer.IsRunning()) m_scroll_timer.Stop();
+}
+
+void TreeCtrl::resetDragItems(bool initSize)
+{
+  if(m_scroll_timer.IsRunning()) m_scroll_timer.Stop();
+  if(m_collapse_timer.IsRunning()) m_collapse_timer.Stop();
+  m_last_mice_item_in_drag_and_drop = NULL;
+  if(initSize) {
+    int width, height;
+    GetSize(&width,  &height);
+    m_lower_scroll_limit = height - GetCharHeight();
+    m_upper_scroll_limit = GetCharHeight();
+  }
+}
+
+void TreeCtrl::markDragItem(const wxTreeItemId itemSrc, bool markIt)
+{
+  if(markIt) {
+    m_drag_text_colour = GetItemTextColour(itemSrc);
+    m_drag_background_colour = GetItemBackgroundColour(itemSrc);
+    
+    SetItemTextColour(itemSrc, *wxWHITE);
+    SetItemBackgroundColour(itemSrc, *wxBLUE);
+  }
+  else {
+    SetItemTextColour(itemSrc, m_drag_text_colour);
+    SetItemBackgroundColour(itemSrc, m_drag_background_colour);
+  }
+}
+
+void TreeCtrl::OnEndDrag(wxTreeEvent& evt)
+{
+  bool makeCopy = ::wxGetKeyState(WXK_CONTROL);
+  bool doOverride = ::wxGetKeyState(WXK_SHIFT);
+  int width, height;
+  wxPoint mousePt = ScreenToClient(wxGetMousePosition());  // Find where the mouse is in relation to the (exited) pane
+  GetSize(&width, &height);  // Store window dimensions
+  int flags = wxTREE_HITTEST_ONITEMICON|wxTREE_HITTEST_ONITEMLABEL;
+  wxTreeItemId currentItem = DoTreeHitTest(mousePt, flags); // Check current before removing root from screen
+  resetDragItems();
+  // Restore Hiden Root
+  SetWindowStyle(m_style);
+  markDragItem(m_drag_item, false);
+  if(m_drag_image) {
+    m_drag_image->Hide();
+    CaptureMouse(); // EndDrag() will release the Mouse, but before call the mouse had been released from event handler
+    m_drag_image->EndDrag();
+    wxDELETE(m_drag_image);
+  }
+  Refresh();
+  
+  wxTreeItemId selectedItem = GetSelection();
+  if(selectedItem.IsOk()) UnselectAll();
+  
+  if((mousePt.x < 0) || (mousePt.x >= width) || (mousePt.y < 0) || (mousePt.y >= height)) { // Not in same area
+    evt.Skip();
+    m_drag_item = NULL;
+    return;
+  }
+  
+  if (m_drag_item != NULL) { // Drag and drop started
+
+    wxTreeItemId itemDst = evt.GetItem();
+    
+    if(itemDst.IsOk() && (currentItem != itemDst)) { // Mice is not over destination item, skip dragging
+      evt.Skip();
+      m_drag_item = NULL;
+      return;
+    }
+
+    if(itemDst && itemDst.IsOk() &&
+       (itemDst != m_drag_item) &&
+       (itemDst != GetItemParent(m_drag_item)) &&
+       ((GetRootItem() != itemDst) || (GetRootItem() != GetItemParent(m_drag_item)))) { // Do not Drag and Drop on its own
+      if(GetRootItem() != itemDst) {
+        if(! ItemIsGroup(itemDst)) {                   // Only group may be destination
+          wxMessageBox(_("Destination is no group"), _("Drag and Drop failed"), wxOK|wxICON_ERROR);
```

Comment:
I would suggest "Destination is not a group.".

Reply:
Proposal is fine with me.